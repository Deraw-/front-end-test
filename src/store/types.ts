// tslint:disable-next-line:no-empty-interface
export interface RootState {}

export interface Product extends Object {
	id: number;
	quantity: number;
}

export interface CartProduct extends Product {
	name: string;
	price: number;
	image: string;
}

// This is used to properly type components using mappers from Vuex
// see https://github.com/vuejs/vuex/pull/1121
// tslint:disable-next-line: no-empty-interface
export default interface VuexBindings {
	cartProducts: CartProduct[];
}
