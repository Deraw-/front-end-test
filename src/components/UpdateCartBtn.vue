<template>
	<div>
		<template v-if="getProductQuantity(product.id)">
			<button
				class="round-btn outline"
				:aria-label="$t('components.product.removeBtn', [product.name])"
				@click="removeFromCart(product)"
			>
				<MinusIcon />
			</button>

			<span class="quantity">
				<span class="sr-only">
					{{ $t('components.product.quantity') }}
				</span>

				{{ getProductQuantity(product.id) }}
			</span>
		</template>

		<button
			class="round-btn"
			:aria-label="$t('components.product.addBtn', [product.name])"
			@click="addToCart(product)"
		>
			<PlusIcon />
		</button>
	</div>
</template>

<script lang="ts">
	import Vue from 'vue';
	import { mapActions, mapGetters, mapState } from 'vuex';

	import MinusIcon from 'icons/Minus.vue';
	import PlusIcon from 'icons/Plus.vue';

	export default Vue.extend({
		name: 'UpdateCartBtn',
		components: {
			MinusIcon,
			PlusIcon
		},
		props: {
			product: {
				type: [Object, Array],
				required: true
			}
		},
		computed: {
			...mapState('cart', ['added'])
		},
		methods: {
			...mapActions('cart', ['addToCart', 'removeFromCart']),
			getProductQuantity(id: number) {
				const product = this.added.find((p: any) => p.id === id);
				const quantity = product ? product.quantity : 0;

				return quantity;
			}
		}
	});
</script>

<style lang="scss" scoped>
	.round-btn {
		width: 35px;
		height: 35px;

		font-size: 1.25rem;

		border-radius: 50%;
		border: 1px solid transparent;

		color: #fff;
		background: #008393;

		&:hover,
		&:focus {
			background: rgb(2, 152, 172);
		}

		&.outline {
			background: transparent;
			border-color: currentColor;
			color: #008393;

			&:hover,
			&:focus {
				background: rgba(2, 152, 172, .1);
			}
		}
	}

	.quantity {
		margin: 0 10px;
	}
</style>
