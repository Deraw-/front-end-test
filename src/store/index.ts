import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import VuexPersistence from 'vuex-persist';
import { RootState } from './types';

// Modules
import cart from './modules/cart';
import products from './modules/products';

// The store is saved in the browser's session
const vuexLocal = new VuexPersistence({
	storage: window.sessionStorage
});

Vue.use(Vuex);

// See https://vuex.vuejs.org/fr/getting-started.html for help
const store: StoreOptions<RootState> = {
	strict: true,
	state: {},
	// See https://vuex.vuejs.org/guide/modules.html for more info on modules
	modules: {
		cart,
		products
	},
	plugins: [
		vuexLocal.plugin
	]
};

export default new Vuex.Store<RootState>(store);
