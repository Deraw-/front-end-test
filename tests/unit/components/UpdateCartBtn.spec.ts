import { shallowMount } from '@vue/test-utils';
import UpdateCartBtn from '@/components/UpdateCartBtn.vue';

import localVue from 'tests-unit/index';
import store from '@/store/';

// Data
import { ProductType } from '@/store/types';
const product: ProductType = {
	id: 0,
	quantity: 15
};

// Build component before each test
let wrapper: any;
beforeEach(() => {
	wrapper = shallowMount(UpdateCartBtn, {
		localVue,
		store,
		propsData: {
			product
		}
	});
});

// Tests
describe('UpdateCartBtn.vue', () => {
	it('renders correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
