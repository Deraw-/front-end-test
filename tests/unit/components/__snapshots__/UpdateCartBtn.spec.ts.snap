// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateCartBtn.vue renders correctly 1`] = `
<div>
  <!----> <button aria-label="components.product.addBtn" class="round-btn">
    <plusicon-stub title="Plus icon" fillcolor="currentColor" size="24"></plusicon-stub>
  </button></div>
`;
